src/runtime-rs/src/shim/src/lib.rs

```diff
@@ -0,0 +1,311 @@
+        }
+
+        // Ensure `address` is a valid path.
+        let path = PathBuf::from(self.address.clone())
+            .canonicalize()
+            .map_err(|_| Error::InvalidArgument)?;
+        let md = path.metadata().map_err(|_| Error::InvalidArgument)?;
+        if !md.file_type().is_socket() {
+            return Err(Error::InvalidArgument);
+        }
+        self.address = path
+            .to_str()
+            .map(|v| v.to_owned())
+            .ok_or(Error::InvalidArgument)?;
+
+        // Ensure `bundle` is a valid path.
+        if !self.bundle.is_empty() {
+            let path = PathBuf::from(self.bundle.clone())
+                .canonicalize()
+                .map_err(|_| Error::InvalidArgument)?;
+            let md = path.metadata().map_err(|_| Error::InvalidArgument)?;
+            if !md.is_dir() {
+                return Err(Error::InvalidArgument);
+            }
+            self.bundle = path
+                .to_str()
+                .map(|v| v.to_owned())
+                .ok_or(Error::InvalidArgument)?;
+        }
+
+        Ok(())
+    }
+
+    fn is_component_dangerous(comp: &str) -> bool {
+        comp.contains(' ') || comp.contains('/') || comp.contains('\t')
+    }
+}
+
+/// Command executor for shim.
+pub struct ShimExecutor {
+    args: ShimArgs,
+}
+
+impl ShimExecutor {
+    /// Create a new instance of [`Shim`].
+    pub fn new(args: ShimArgs) -> Self {
+        ShimExecutor { args }
+    }
+
+    fn load_image_spec(&self) -> Result<oci_spec::runtime::Spec> {
+        let bundle_path = self.get_bundle_path()?;
+        let spec_file = bundle_path.join("config.json");
+
+        oci_spec::runtime::Spec::load(spec_file).map_err(Error::OciSpec)
+    }
+
+    fn write_address(&self, address: &str) -> Result<()> {
+        let dir = self.get_bundle_path()?;
+        let file_path = &dir.join("address");
+        std::fs::write(file_path, address.as_bytes())
+            .map_err(|e| Error::WriteFile(e, file_path.clone()))
+    }
+
+    fn write_pid_file(&self, pid: &str) -> Result<()> {
+        let dir = self.get_bundle_path()?;
+        let file_path = &dir.join("shim.pid");
+        std::fs::write(file_path, pid.as_bytes())
+            .map_err(|e| Error::WriteFile(e, file_path.clone()))
+    }
+
+    fn read_pid_file(&self, bundle_path: &Path) -> Result<String> {
+        let file_path = bundle_path.join("shim.pid");
+        // Limit the size of content to read in.
+        let size = file_path.metadata().unwrap().len();
+        if size > 32 {
+            return Err(Error::InvalidArgument);
+        }
+
+        std::fs::read_to_string(&file_path).map_err(|e| Error::OpenFile(e, file_path.clone()))
+    }
+
+    fn get_bundle_path(&self) -> Result<PathBuf> {
+        std::env::current_dir().map_err(Error::BundlePath)
+    }
+
+    fn socket_address(&self, id: &str) -> String {
+        format!(
+            "{}/{}/{}/{}",
+            SOCKET_ROOT, &self.args.namespace, id, "shim.sock"
```

Comment:
`shim.sock` should be a global `const` to make it more obvious what's being hard-coded, and to help the unit tests.

Reply:
Fixed in the first revision